ccp/lib/clients/base.py

```diff
@@ -23,3 +25,57 @@ def __init__(self, base_command):
         """
         super(Client, self).__init__()
         self.base_command = base_command
+
+
+class APIClient(Client):
+    """
+    Base class for all clients that query an API Server for information
+    """
+    def __init__(
+            self,
+            server,
+            secure=True,
+            verify_ssl=True,
+            authorization=None
+    ):
+        """
+        Initialize APIClient
+        :param server: The URL/IP server to hit.
+        :type server str
+        :param secure: Default True: Use SSL for queries.
+        :type secure bool
+        :param verify_ssl: Default True: Verify SSL certificate.
+        :param authorization: Set if you want or need to use Authorization
+        :type authorization Authorization
+        """
+        self.server_endpoint = "{}://{}".format(
+            "https" if secure else "http",
+            server if server else "localhost"
+        )
+        self.authorization = authorization
+        self.skip_ssl = verify_ssl
+
+    def _query(self, target, headers=None):
+        """
+        Given a target (query without url), queries endpoint and gets data
+        :param target: The query to send to the api server, after the hostname
+        and port. For eg: URL = http://example.com/hello, then target will be
+        '/hello'
+        :type target str
+        :param headers: Any headers you wish to add to query.
+        :type headers dict
+        :return:The response object and exception object, if any was raised.
+        """
+        query_url = "{}{}".format(
```

Comment:
mind checking if `target` starts with  `/` OR `self.server_endpoint` ends with `/`.
This is a common problem and that's why library utils like `urljoin` is devised.

Please use `urljoin` here to combine, target with server end point.

I think its a must.

Reply:
I'll request change for this.

We are going to have a lot of places in our code where we need to perform this, and lets make sure we do that via `urljoin` and leave no room for any input validation errors!

Reply:
Moved to urljoin